### ISSUE TITLE
Confused about SIMD blogpost

### ISSUE THREAD

**Issue by Narsil** [matklad/matklad.github.io#121]
I tried the last function from 

https://matklad.github.io/2023/04/09/can-you-trust-a-compiler-to-optimize-your-code.html

and got 

https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:1,endLineNumber:7,positionColumn:1,positionLineNumber:7,selectionStartColumn:1,selectionStartLineNumber:7,startColumn:1,startLineNumber:7),source:'use+std::iter::zip%3B%0A%0Apub+fn+commone_prefix5(xs:+%26%5Bu8%5D,+ys:+%26%5Bu8%5D)+-%3E+usize+%7B%0A%0A++let+chunk_size+%3D+16%3B%0A%0A%0A++let+off+%3D%0A%0A++++zip(xs.chunks_exact(chunk_size),+ys.chunks_exact(chunk_size))%0A%0A++++++.take_while(%7C(xs_chunk,+ys_chunk)%7C+xs_chunk+%3D%3D+ys_chunk)%0A%0A++++++.count()+*+chunk_size%3B%0A%0A%0A++off+%2B+zip(%26xs%5Boff..%5D,+%26ys%5Boff..%5D)%0A%0A++++.take_while(%7C(x,+y)%7C+x+%3D%3D+y)%0A%0A++++.count()%0A%0A%7D'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1680,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'-C+target-feature%3D%2Bavx2',selection:(endColumn:37,endLineNumber:6,positionColumn:37,positionLineNumber:6,selectionStartColumn:37,selectionStartLineNumber:6,startColumn:37,startLineNumber:6),source:1),l:'5',n:'0',o:'+rustc+1.68.0+(Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4

It doesn't seem like SIMD is kicking in like it should.

Am I missing something ?
I'd like to move to category 3 :)

Edit: I tried changing  the size to chunked = 4 and `f32` instead of `u8` to help it move to `__mm128` but to no avail.

**Comment by matklad** [comment]
you are missing  `-C opt-level=3`

**Comment by Narsil** [comment]
Of course. Thanks !

**Comment by Narsil** [comment]
I find it hard to make it output packed SIMD instead of singular. Any other tricks ?


https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:1,endLineNumber:7,positionColumn:1,positionLineNumber:7,selectionStartColumn:1,selectionStartLineNumber:7,startColumn:1,startLineNumber:7),source:'use+std::iter::zip%3B%0A%0Apub+fn+commone_prefix5(xs:+%26%5Bu8%5D,+ys:+%26%5Bu8%5D)+-%3E+usize+%7B%0A%0A++let+chunk_size+%3D+16%3B%0A%0A%0A++let+off+%3D%0A%0A++++zip(xs.chunks_exact(chunk_size),+ys.chunks_exact(chunk_size))%0A%0A++++++.take_while(%7C(xs_chunk,+ys_chunk)%7C+xs_chunk+%3D%3D+ys_chunk)%0A%0A++++++.count()+*+chunk_size%3B%0A%0A%0A++off+%2B+zip(%26xs%5Boff..%5D,+%26ys%5Boff..%5D)%0A%0A++++.take_while(%7C(x,+y)%7C+x+%3D%3D+y)%0A%0A++++.count()%0A%0A%7D'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1680,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'-C+target-feature%3D%2Bavx2',selection:(endColumn:37,endLineNumber:6,positionColumn:37,positionLineNumber:6,selectionStartColumn:37,selectionStartLineNumber:6,startColumn:37,startLineNumber:6),source:1),l:'5',n:'0',o:'+rustc+1.68.0+(Editor+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4

**Comment by matklad** [comment]
I see that the addition is packed (`vaddps`), but the moves are scalar (`vmovsd`). I _think_ this is because packed moves require that memory is aligned. So the compiler can not just generate a single packed move here. Why it doesn't issue a bunch of scalar moves and then a bunch of packing instructions is unclear, but presumably that might be not worth it over just loading and summing as we go?

So, my hypothesis is that, to get packed operations here, it is also nesesary to align the slice and make it visible in the type system that the slice is suitably aligned. 

https://doc.rust-lang.org/stable/std/primitive.slice.html#method.align_to can potentially help here.

In other words, I think there are two aspects for vector code:

* iteration in batches
* aligned batches

the `.chunks_exact` trick explains the fist bit to the compiler, but in this case we also need the second part
